### ISSUE TITLE
Skipped scenarios are executed if Before hook is not the last one

### ISSUE THREAD

**Issue by lashchev** [cucumber/cucumber-js#976]
(v3.1.0)
If I create 2 hooks and first hook returns "skipped" but there is another "Before" hook defined after it - then scenario will be executed anyway.

For example:
```
Before({ tags: "@NotImplemented" }, function (scenario) {
    return 'skipped';
});

Before(function (scenario) {
    return new Promise((resolve, reject) => {
        resolve();
    });
});

 @NotImplemented
  Scenario: Some feature that is not yet implemented
```

=>>> scenario will be executed

But if I rearrange hooks in the file - scenario will not be executed.

Also, if one hook returned "skipped" - all further hooks should not be executed.

**Comment by charlierudolph** [comment]
Hmm, this is an interesting idea. Currently even if a before hook fails, other before hooks still run and all the after hooks run. 

We can update it so that for before hooks if one skipped / failed, then we skip the rest of the before hooks but I still think all after hooks should still run in order to teardown anything setup by the before hooks. 

Thus after hooks need to be implemented in a way that they guard against the matching before hook not running (in the case was skipped)

What do you think?

**Comment by charlierudolph** [comment]
Closing as this was fixed in #1056

**Comment by Aksahy-sh** [comment]
Failure in before hook:Hooks.setup(Scenario)
Message: java.lang.NullPointerException

Why do we get this kind of errors
